Civi/Api4/Generic/Traits/DAOActionTrait.php

```diff
@@ -118,6 +118,9 @@ protected function writeObjects($items) {
 
     foreach ($items as &$item) {
       $entityId = $item[$idField] ?? NULL;
+      if (CoreUtil::isContact($this->getEntityName())) {
+        $entityId = FormattingUtil::resolveContactID($idField, $entityId);
+      }
```

Comment:
entityID at this point might still be "user_contact_id". That doesn't matter for the main contact save but it does for the call to `formatCustomParams` which ends up writing an SQL query via singleValueQuery with `WHERE id=user_contact_id`.